core/src/main/java/org/elasticsearch/index/fielddata/AbstractScriptDocValues.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.fielddata;
+
+import java.util.AbstractList;
+import java.util.Comparator;
+import java.util.function.UnaryOperator;
+
+/**
+ * Implements {@link AbstractList} to throw helpful error messages when someone attempts to modify the doc values list.
+ */
+public abstract class AbstractScriptDocValues<T> extends AbstractList<T> {
```

Comment:
Since `ScriptDocValues` is an interface this can't be a private inner class. I can only be public. I could have made this private or package private but I didn't because I figured implementers may want to use it. We do let plugins add mappers.

It might make more sense to make `ScriptDocValues` an abstract class and add these?